plugins/package-managers/node/src/main/kotlin/Yarn.kt

```diff
@@ -90,12 +90,14 @@ class Yarn(
     override fun getRemotePackageDetails(workingDir: File, packageName: String): PackageJson? {
         yarnInfoCache.read(packageName)?.let { return parsePackageJson(it) }
 
-        val process = run(workingDir, "info", "--json", packageName)
-
-        return parseYarnInfo(process.stdout)?.also {
-            yarnInfoCache.write(packageName, Json.encodeToString(it))
-        } ?: checkNotNull(parseYarnInfo(process.stderr)).also {
-            logger.warn { "Error running '${process.commandLine}' in directory $workingDir: $it" }
+        return runCatching {
```

Comment:
> While at it, stop to attempt parsing `stderr` to a `PackageJson` instance, because `stderr` won't contain that.

I don't think that's correct. For example `yarn info --json foo.txt > /dev/null` (which suppresses stdout to only see stderr) returns `{"type":"error","data":"Received invalid response from npm."}`.

Reply:
> That's not correct. `run()` just returns a `ProcessCapture` instance and does not throw on a non-zero exit code unless also `requireSuccess()` is called.

This is what I've been assuming as well when starting with this PR, until I came across this `requireSuccess()` call here, which left me surprised: https://github.com/oss-review-toolkit/ort/blob/131c1308cd1934806c2e534cba0496c27402a958/utils/common/src/main/kotlin/CommandLineTool.kt#L70-L76

Reply:
> > While at it, stop to attempt parsing `stderr` to a `PackageJson` instance, because `stderr` won't contain that.
> 
> I don't think that's correct. For example `yarn info --json foo.txt > /dev/null` (which suppresses stdout to only see stderr) returns `{"type":"error","data":"Received invalid response from npm."}`.

The JSON returned is not in 'package.json' format. Can you explain why you think it's not correct?

Reply:
> This is what I've been assuming as well when starting with this PR, until I came across this `requireSuccess()` call here, which left me surprised:

Sorry, I mixed things up: `run()` is a short-hand for constructing a `ProcessCapture()` and calling `requireSuccess()`. If you don't want to throw in case of a non-zero exit code, you should simply call the `ProcessCapture()` constructor instead of `run()`, and probably not even implement `CommandLineTool`. But the latter is a bit difficult currently with `Npm` already implementing `CommandLineTool`. So all in all, while implementation-wise a bit involved, I agree that wrapping `run()` in `runCatching` is convenient at least from the caller perspective.

Reply:
Do you still think the PR needs to be re-worked? How shall we proceed ?

Reply:
> The JSON returned is not in 'package.json' format, wich is what I've been trying to say.

Now I see what you mean... I believe this is something that git broken in ad9a3639604aacfa6aa3fd831593201f42e78483. Before that commit, parsing was performed with generic JSON elements, and an error message was extracted from the `data` object. After that commit, even error messages are sent through `parsePackageJson` which then throws and `parseYarnInfo()` returns null instead of an error message. That is, currently the `logger.warn {}` would never be called, as the `checkNotNull()` before would already throw... I believe this could be fixed by not calling `parsePackageJson()` at all as part of `parseYarnInfo()` (which would funnily also better match its docs) and changing the code in the second commit to this:

    override fun getRemotePackageDetails(workingDir: File, packageName: String): PackageJson? {
        yarnInfoCache.read(packageName)?.let { return parsePackageJson(it) }

        val process = run(workingDir, "info", "--json", packageName)

        return parseYarnInfo(process.stdout)?.let { parsePackageJson(it) }?.also {
            yarnInfoCache.write(packageName, Json.encodeToString(it))
        } ?: parseYarnInfo(process.stderr)?.let {
            logger.warn { "Error running '${process.commandLine}' in directory $workingDir: ${it.jsonPrimitive.content}" }
            null
        }
    }

```
internal fun parseYarnInfo(output: String): JsonElement? =
    runCatching {
        output.byteInputStream().use { inputStream ->
            Json.decodeToSequence<JsonObject>(inputStream)
                .firstOrNull { (it["type"] as? JsonPrimitive)?.content == "inspect" }?.let {
                    it["data"]
                }
        }
    }.getOrNull()
```

Reply:
I still have doubts that the error handling code path is used at all. 
We don't have any test using it, and more importantly, when the call to run does not throw the command has been executed successfully. Sucessful to me means, one can rely on the result. And if parsing the `stdout` of a successful execution was null, then the registry doesn't know about the package, which is ok.
Examining stderr, whether there is an error we could log, and then say that command execution failed, even though it succeeded seems odd to me. 

Can we just strip that code (error handling for successful exit code), and re-introduce it once we should have evidence that it is actually needed?

Reply:
This is where the code comes from: https://github.com/oss-review-toolkit/ort/commit/fb7cdabb667492163ac380da142836005d2bf4c1

Reply:
Ok, if reproduced it. The code is indeed reachable:
```
 yarn info --json abddeac@1.0.0 > /dev/null
{"type":"warning","data":"package.json: \"test\" is also the name of a node core module"}
{"type":"warning","data":"package.json: No license field"}
{"type":"error","data":"Received invalid response from npm."}
#:~$ echo $?
0
```

Reply:
@sschuberth I decided to make a separate PR to address the issues we found here in the info parsing. 
Let's process this one first, once done I'll update / rebase this PR on top. See https://github.com/oss-review-toolkit/ort/pull/9323.

Reply:
> I still have doubts that the error handling code path is used at all.

...

> Ok, i've reproduced it. The code is indeed reachable:

Yes, I also remember to have run into it in real life.

Reply:
I believe this topics has now been fully addressed by https://github.com/oss-review-toolkit/ort/pull/9328.